File: src/main/java/com/mikea/bayes/query/QueryAlgorithm.java
package com.mikea.bayes.query;

import com.mikea.bayes.BayesianNetwork;
import com.mikea.bayes.Factor;
import com.mikea.bayes.SumProduct;
import com.mikea.bayes.VarAssignment;
import com.mikea.bayes.VarSet;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * @author mike.aizatsky@gmail.com
 */
public interface QueryAlgorithm<R extends QueryAlgorithm.Result> {
    VarElimination VAR_ELIMINATION = new VarElimination(new SumProduct.MinNeighborsStrategy());
    QueryAlgorithm DEFAULT = VAR_ELIMINATION;

    R prepare(BayesianNetwork network);

    abstract class Result {
        public abstract Factor query(VarSet query, @Nullable VarAssignment evidence);

        public double getProbability(@Nonnull VarAssignment evidence) {
            Factor factor = query(VarSet.newVarSet(evidence.getVars()), null);
            return factor.getValue(evidence.getVars(), evidence.getValues());
        }

        public Factor query(VarSet query) {
            return query(query, null);
        }
    }
}

File: src/main/java/com/mikea/bayes/query/QueryAlgorithm.scala
package com.mikea.bayes.query

import com.mikea.bayes.BayesianNetwork
import com.mikea.bayes.Factor
import com.mikea.bayes.SumProduct
import com.mikea.bayes.VarAssignment
import com.mikea.bayes.VarSet
//remove if not needed

object QueryAlgorithm {
  var VAR_ELIMINATION: VarElimination = new VarElimination(new SumProduct.MinNeighborsStrategy())
  var DEFAULT = VAR_ELIMINATION

  abstract class Result {

    def query(query: VarSet, evidence: VarAssignment): Factor

    def getProbability(evidence: VarAssignment): Double = ???/*{
      val factor = query(VarSet.newVarSet(evidence.getVars), null)
      factor.getValue(evidence.getVars, evidence.getValues)
    }*/

    def query(q: VarSet): Factor = query(q, null)
  }
}
/**
 * @author mike.aizatsky@gmail.com
 */
trait QueryAlgorithm[R <: QueryAlgorithm.Result] {
  def prepare(network: BayesianNetwork): R
}
